packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/EditorToolbarDecorator.kt

```diff
@@ -1,233 +1,142 @@
-    internal var inferredDatabase: String? = null
-    internal var guessedDialect: Dialect<PsiElement, Project>? = null
-    internal lateinit var toolbar: MdbJavaEditorToolbar
-    internal lateinit var editor: Editor
-    internal lateinit var messageBusConnection: MessageBusConnection
-
-    fun onConnected(dataSource: LocalDataSource) {
-        editor.virtualFile?.putUserData(Keys.attachedDataSource, dataSource)
-        editor.virtualFile?.removeUserData(Keys.attachedDatabase)
-
-        ApplicationManager.getApplication().invokeLater {
-            val project = editor.project!!
-            val session = DatabaseSessionManager.openSession(project, dataSource, null)
-            val probe = NewConnectionActivatedProbe()
-            probe.connected(session)
-
-            inferredDatabase?.let {
-                toolbar.databaseComboBox.selectedDatabase = inferredDatabase
-            }
-
-            analyzeFileFromScratch()
+    JdbcDriverManager.Listener {
+    private lateinit var projectService: ProjectService
+    private lateinit var dataSourceModel: DataSourceModel
+    private lateinit var databaseModel: DatabaseModel
+    private lateinit var toolbar: MdbJavaEditorToolbar
+
+    private fun Editor.toggleMdbJavaEditorToolbar() {
+        if (projectService.selectedEditor == null) {
+            toolbar.detachFromEditor(this)
+        } else {
+            toolbar.attachToEditor(this, projectService.databaseComboBoxVisibleForCurrentEditor())
         }
     }
 
-    fun onDisconnected() {
-        editor.virtualFile?.removeUserData(Keys.attachedDataSource)
-        editor.virtualFile?.removeUserData(Keys.attachedDatabase)
-        analyzeFileFromScratch()
-    }
-
-    fun onDatabaseSelected(database: String) {
-        editor.virtualFile?.putUserData(Keys.attachedDatabase, database)
-        analyzeFileFromScratch()
+    private fun setupSubscriptionsForProject(project: Project) {
+        val messageBusConnection = project.messageBus.connect()
+        messageBusConnection.subscribe(FileEditorManagerListener.FILE_EDITOR_MANAGER, this)
+        messageBusConnection.subscribe(PsiModificationTracker.TOPIC, this)
+        messageBusConnection.subscribe(DataSourceManager.TOPIC, this)
+        messageBusConnection.subscribe(JdbcDriverManager.TOPIC, this)
     }
 
-    fun onDatabaseUnselected() {
-        editor.virtualFile?.removeUserData(Keys.attachedDatabase)
-        analyzeFileFromScratch()
-    }
-
-    override fun selectionChanged(event: FileEditorManagerEvent) {
-        (event.newEditor as? TextEditor)?.editor?.let { newEditor ->
-            editor = newEditor
-            editor.project?.let { project ->
-                guessAndStoreDialect(project)
-            }
-            ensureToolbarIsVisibleIfNecessary()
-            toolbar.reloadDatabases()
+    private fun runReadActionWhenInitialised(block: () -> Unit) {
+        ApplicationManager.getApplication().runReadAction {
+            block()
```

Comment:
Probably forgot to rename this function to something else :D

Reply:
yea I did not include the logic to actually check wether the activity started. Have done that now 👍